common/fs/src/cache/mod.rs

```diff
@@ -199,66 +214,36 @@ where
+                    .map(|entry| self.entry_path_passes(entry, from_name.clone(), &_entries))
+                    .unwrap_or(false);
 
-                        let mut path = self.resolve_direct_path(
-                            &_entries.get(entries[0]).unwrap().borrow(),
-                            &_entries,
-                        );
-                        path.push(from_name.clone());
-                        path
-                    }
-                    None => {
-                        error!(
-                            "got move event where from is an untracked watch descriptor: {:?}",
-                            from_wd
-                        );
-                        return;
-                    }
-                };
-
-                let to_path = match self.watch_descriptors.get(&to_wd) {
-                    Some(entries) => {
-                        if entries.is_empty() {
-                            error!("got move event where to watch descriptors maps to no entries: {:?}", to_wd);
-                            return;
-                        }
-
-                        let mut path = self.resolve_direct_path(
-                            &_entries.get(entries[0]).unwrap().borrow(),
-                            &_entries,
-                        );
-                        path.push(to_name.clone());
-                        path
-                    }
-                    None => {
-                        error!(
-                            "got move event where to is an untracked watch descriptor: {:?}",
-                            to_wd
-                        );
-                        return;
-                    }
-                };
-
-                let is_to_path_ok = self.passes(&to_path, &_entries);
-
-                let is_from_path_ok = self.passes(&from_path, &_entries);
+                let is_to_path_ok = self
+                    .get_first_entry(&to_wd)
+                    .map(|entry| self.entry_path_passes(entry, to_name.clone(), &_entries))
+                    .unwrap_or(false);
 
                 if is_to_path_ok && is_from_path_ok {
-                    self.process_move(&from_wd, from_name, &to_wd, to_name, events, &mut _entries);
+                    self.process_move(&from_wd, from_name, &to_wd, to_name, events, &mut _entries)
                 } else if is_to_path_ok {
-                    self.process_create(&to_wd, to_name, events, &mut _entries);
+                    self.process_create(&to_wd, to_name, events, &mut _entries)
                 } else if is_from_path_ok {
-                    self.process_delete(&from_wd, from_name, events, &mut _entries);
+                    self.process_delete(&from_wd, from_name, events, &mut _entries)
+                } else {
+                    Err("Moving paths are invalid".into())
                 }
             }
-            WatchEvent::Overflow => panic!("overflowed kernel queue!"),
+            WatchEvent::Overflow => Err(
```

Comment:
Hmm, is this recoverable? I'm not actually sure of the implications, maybe we should dump the fs cache and repopulate?

I think panic'ing is sensible here until we handle it more gracefully as the pod restart should get things back into a clean state.

Reply:
I'm also in doubt of what to do in this situation...
Queue overflow might indicate a third party system that's misbehaving and `fsync`ing lots of small changes in lots of files or a ridiculously low value in `/proc/sys/fs/inotify/max_queued_events`.

I think that logging at error level acknowledging the issue is less aggressive than exiting the process. We could add improvements like sending an ERROR log message to logdna servers for operators to easily pick up on them.

wdyt? Should I create a ticket for an improvement? Should I revert to panic?

Reply:
We havn't had any reports of this panic happening, maybe we could create the err, log it and then immediately .expect it as an interim measure until we handle it properly?

I think handling it properly involves recreating the fs cache and the tailers from the last known good position which will be possible when I get the stateful lookback set up.

Reply:
> We haven't had any reports of this panic happening

I agree that this is a rare edge case.

> .expect it as an interim measure until we handle it properly

ok, I'll update the patch to log at error level and panic.